src/Microsoft.VisualStudio.Threading/ReentrantSemaphore.cs

```diff
@@ -0,0 +1,461 @@
+            }
+        }
+
+        /// <summary>
+        /// Executes a given operation within the semaphore.
+        /// </summary>
+        /// <param name="operation">The delegate to invoke once the semaphore is entered.</param>
+        /// <param name="cancellationToken">A cancellation token.</param>
+        /// <returns>A task that completes with the result of <paramref name="operation"/>, after the semaphore has been exited.</returns>
+        public abstract Task ExecuteAsync(Func<Task> operation, CancellationToken cancellationToken = default);
+
+        /// <inheritdoc/>
+        public void Dispose()
+        {
+            this.Dispose(true);
+            GC.SuppressFinalize(this);
+        }
+
+        /// <summary>
+        /// Disposes managed and unmanaged resources held by this instance.
+        /// </summary>
+        /// <param name="disposing"><c>true</c> if <see cref="Dispose()"/> was called; <c>false</c> if the object is being finalized.</param>
+        protected virtual void Dispose(bool disposing)
+        {
+            if (disposing)
+            {
+                this.semaphore.Dispose();
+            }
+        }
+
+        /// <summary>
+        /// Throws an exception if this instance has been faulted.
+        /// </summary>
+        protected virtual void ThrowIfFaulted()
+        {
+        }
+
+        /// <summary>
+        /// Disposes the specfied release, swallowing certain exceptions.
+        /// </summary>
+        /// <param name="releaser">The releaser to dispose.</param>
+        private static void DisposeReleaserNoThrow(AsyncSemaphore.Releaser releaser)
+        {
+            try
+            {
+                releaser.Dispose();
+            }
+            catch (ObjectDisposedException)
+            {
+                // Swallow this, since in releasing the semaphore if it's already disposed the caller probably doesn't care.
+            }
+        }
+
+        /// <summary>
+        /// An implementation of <see cref="ReentrantSemaphore"/> supporting the <see cref="ReentrancyMode.NotRecognized"/> mode.
+        /// </summary>
+        private class NotRecognizedSemaphore : ReentrantSemaphore
+        {
+            /// <summary>
+            /// Initializes a new instance of the <see cref="NotRecognizedSemaphore"/> class.
+            /// </summary>
+            /// <param name="initialCount">The initial number of concurrent operations to allow.</param>
+            /// <param name="joinableTaskContext">The <see cref="JoinableTaskContext"/> to use to mitigate deadlocks.</param>
+            internal NotRecognizedSemaphore(int initialCount, JoinableTaskContext joinableTaskContext)
+                : base(initialCount, joinableTaskContext)
+            {
+            }
+
+            /// <inheritdoc />
+            public override async Task ExecuteAsync(Func<Task> operation, CancellationToken cancellationToken = default)
```

Comment:
>ExecuteAsync [](start = 39, length = 12)

Style question:  It looks like prolog and epilog of each concrete class's implementation of ```ExecuteAsync``` are identical.  Is it worth putting ```ExecuteAsync``` in the base class and create an abstract method to return the right ```semaphoreUser``` delegate for each concrete type?

Doing it this way might make it easier to add overloads of ```ExecuteAsync``` if you choose do so in the future.

Reply:
Done.